src/anomalib/models/image/uflow/nfa_tree.py

```diff
@@ -0,0 +1,221 @@
+"""Description:
+    This module contains the implementation of the NFA tree algorithm, used in the U-Flow model to compute the
+    log-probability map from the latent variables, and then generate a mask, performing an automatic segmentation of
+    anomalies.
+    We construct a tree of upper level sets from the latent variables, and compute the log-probability of each region in
+    the image. We then perform a prune and merge process to remove nodes with high PFA values. The final tree is used to
+    compute the log-NFA, which is used to obtain the segmentation mask with an automatic threshold.
+
+Note:
+    This code is part of the U-Flow model, and is not used in the basic functionality of training and testing. It is
+    included as it is part of the U-Flow paper, and can be called separately if an unsupervised anomaly segmentation is
+    needed.
+
+Licence:
+    Copyright (C) 2022-2024 Intel Corporation
+    SPDX-License-Identifier: Apache-2.0
+"""
+import itertools as it
+from typing import Union
+
+import networkx as nx
+import numpy as np
+import torch
+from mpmath import mp
+from skimage.morphology import max_tree
+
+mp.dps = 15
+
+
+class NFATree:
```

Comment:
Would it be possible to port the np computations to torch in this implementation? We try to minimise the use of numpy and use torch instead.